### PR TITLE
HW4 Gof Patterns

### PR DESCRIPTION
**Pull request by seemsod1**
- implemented pattern Chain of Responsibility to work with rate APIs
- implemented LoggerClient to log responses from rate APIs
- added two new providers: NBU and PrivatBank
- added error wrapping

### CODE REVIEW

**Comment by vladyslavpavlenko** [comment]
Seems good to me 👍

File: .example.env [hunk 1]
```diff
@@ -5,4 +5,4 @@ MAILER_PORT=<port>
 MAILER_FROM=<email>
 MAILER_PASSWORD=<password>
 
-APP_PROD=<bool>
\ No newline at end of file
+APP_PROD={prod|dev|test}
```
Comment (vladpistun) [inline .example.env]: plz rename it to APP_ENV

File: cmd/web/setup.go [hunk 2]
```diff
@@ -22,36 +25,51 @@ func setup(_ *config.AppConfig) error {
...
+		return fmt.Errorf("connecting to database: %w", err)
 	}
 
 	if err = db.RunMigrations(); err != nil {
 		log.Println("Cannot run schemas migration! Dying...")
-		return err
+		return fmt.Errorf("running schemas migration: %w", err)
 	}
-	provider := rateapi.NewCoinbaseProvider()
+	CoinBaseProvider := rateapi.NewLoggingClient("api.coinbase.com",
```
Comment (vladpistun) [inline cmd/web/setup.go]: lets move api_urls to env, it will help us to use sandbox/prod api urls

File: internal/handlers/subscribe.go [hunk 3]
```diff
@@ -14,6 +16,7 @@ import (
 // Subscribe subscribes a user to the newsletter
 func (m *Repository) Subscribe(w http.ResponseWriter, r *http.Request) {
 	if err := r.ParseMultipartForm(10 << 20); err != nil {
+		log.Println(fmt.Errorf("parsing form: %w", err))
```
Comment (vladpistun) [inline internal/handlers/subscribe.go:19]: remove log over ur code

**Review by vladpistun — APPROVED** [review]

File: internal/handlers/subscribe.go (hunk 3 above)
Comment (seemsod1) [reply internal/handlers/subscribe.go:19]: But how do I understand where the error happened, without logging?

**Pull request merged**